r2-core/src/main/java/com/linkedin/r2/transport/http/client/EvictingCircularBuffer.java

```diff
@@ -0,0 +1,245 @@
+/*
+   Copyright (c) 2021 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.r2.transport.http.client;
+
+import com.linkedin.common.callback.Callback;
+import com.linkedin.common.util.None;
+import com.linkedin.r2.transport.http.client.ratelimiter.CallbackStore;
+import com.linkedin.util.clock.Clock;
+import com.linkedin.util.clock.SystemClock;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+
+/**
+ * A CallbackStore specifically designed to feed an asynchronous event loop with a constant supply of unique Callbacks.
+ *
+ * EvictingCircularBuffer accomplishes a few key goals:
+ * - Must always accept submission of new Callbacks, replacing oldest Callbacks when buffer is exhausted
+ * - Must provide a unique Callback between subsequent get() calls
+ * - Must be able to honor get requests in excess of put requests, returning previously returned Callbacks is acceptable.
+ * - During periods of insufficient write throughput, must prune stale Callbacks, ultimately throwing NoSuchElementException
+ *   to upstream callers when inbound write throughput has dropped to zero.
+ * - Must do the above with high performance, without adding meaningful latency to reader/writer threads.
+ *
+ * This class is thread-safe, achieving performance through granular locking of each element of the underlying circular buffer.
+ */
+@SuppressWarnings("serial")
+public class EvictingCircularBuffer implements CallbackStore
+{
+  private Duration _ttl;
+  private final ArrayList<Callback<None>> _callbacks = new ArrayList<>();
+  private final ArrayList<Instant> _ttlBuffer = new ArrayList<>();
+  private final ArrayList<ReentrantReadWriteLock> _elementLocks = new ArrayList<>();
+  private final AtomicInteger _readerPosition = new AtomicInteger();
+  private final AtomicInteger _writerPosition = new AtomicInteger();
+  private final Clock _clock;
+
+
+  public EvictingCircularBuffer(int capacity, int ttl, ChronoUnit ttlUnit)
+  {
+    this(capacity, ttl, ttlUnit, SystemClock.instance());
+  }
+
+  EvictingCircularBuffer(int capacity, int ttl, ChronoUnit ttlUnit, Clock clock)
```

Comment:
Also make this public? Can we add a javadoc here?

Reply:
Opted to remove the constructor that does not have Clock as a parameter. Given that we already need a Clock for SmoothRateLimiter (and ConstantQpsRateLimiter), it makes sense to re-use the Clock passed into there.